### PR TITLE
add_docker

### PR DESCRIPTION
**Pull request by ducanhdt**
in this pull request, i add the docker and docker-compose file so that people can easier to run the code.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,6 +22,11 @@ Get your OpenAI API Key [here](https://platform.openai.com/account/api-keys) and
 pip install codeinterpreterapi
 ```
 
+### Using docker
+```bash
+docker-compose up --build
```
Comment (shroominic) [inline README.md:27]: Why docker-compose? I don't see the need for this for a single container application

Comment (ducanhdt) [reply README.md:27]: you are right, but i am about to attempt to use Redis memory so I think maybe someone will need docker-compose to run memory manager components like Redis or SQL,

Comment (ducanhdt) [reply README.md:27]: I just push another commit which save chat memory via the session id to json file or redis

Comment (shroominic) [reply README.md:27]: Yeah but still I think its not the place to for a docker-compose file. Everyone should do that in their own backend setup.

File: codeinterpreterapi/session.py [hunk 2]
```diff
@@ -9,24 +10,45 @@
...
         self,
+        session_id=None,
         model=None,
         openai_api_key=settings.OPENAI_API_KEY,
         verbose=settings.VERBOSE,
-        tools: list[BaseTool] = None
+        tools: list[BaseTool] = None,
+        history_storage=os.environ['STORAGE_TYPE'],
     ) -> None:
+        if not session_id:
```
Comment (shroominic) [inline codeinterpreterapi/session.py:42]: Codebox already has an id so maybe just a property?

```python
@property
def session_id(self) -> int: return self.codebox.id
```

File: codeinterpreterapi/session.py [hunk 3]
```diff
@@ -80,13 +104,25 @@ def _agent(self) -> BaseSingleActionAgent:
         )
 
     def _agent_executor(self) -> AgentExecutor:
+        if self.history_storage == "file":
+            history = FileChatMessageHistory(
+                f"{os.environ['SESSION_FOLDER_STORAGE']}/{self.session_id}.json"
+            )
+        elif self.history_storage == "redis":
+            history = RedisChatMessageHistory(
```
Comment (shroominic) [inline codeinterpreterapi/session.py:112]: I really like this idea! Maybe this can be done with the option to input your own history backend when creating the session.

File: README.md (hunk 1 above)
Comment (ducanhdt) [reply README.md:27]: It is my habit to use docker-compose to start the system in the most convenient way without remembering docker variables. So in case, you think it is unnecessary, I will remove it in the next commit

File: codeinterpreterapi/session.py (hunk 2 above)
Comment (ducanhdt) [reply codeinterpreterapi/session.py:42]: the purpose of this session_id here is that it could be used to load the old memory of previous chat session.

Comment (ducanhdt) [reply codeinterpreterapi/session.py:42]: I don't know if it possible to restore a codebox with its id

File: codeinterpreterapi/session.py (hunk 3 above)
Comment (ducanhdt) [reply codeinterpreterapi/session.py:112]: yes, it is my target, if the history exists it will be loaded, otherwise, the new empty one will be created. It is convenient to check the problem when an error is occupied.

**Comment by shroominic** [comment]
Hey I really like your memory implementation idea but do not agree with the docker part so can you maybe remove it or split the PR so we can work seperate on these two topics?

**Comment by ducanhdt** [comment]
> Hey I really like your memory implementation idea but do not agree with the docker part so can you maybe remove it or split the PR so we can work seperate on these two topics?

I just removed the docker file like you said

**Comment by ducanhdt** [comment]
Hi @shroominic, I encountered a conflict of GitHub action so that I can't pass black and isort, please help me with it.

**Comment by shroominic** [comment]
Hey @ducanhdt,
yeah sorry I should have mentioned this I implemented a pre-commit script which ensures a good code quality.
To run this you can do `pip install pre-commit` and then in the root of the project do `pre-commit install`.
To maually run it you can do `pre-commit run --all-files`.

**Comment by shroominic** [comment]
I see the only thing missing is just running black this will get automatically fixed when you run the pre-commit script :)

File: codeinterpreterapi/prompts/remove_dl_link.py [hunk 4]
```diff
@@ -1,4 +1,5 @@
-from langchain.prompts.chat import ChatPromptTemplate, HumanMessagePromptTemplate
+from langchain.prompts.chat import (ChatPromptTemplate,
```
Comment (shroominic) [inline codeinterpreterapi/prompts/remove_dl_link.py:1]: why did you change this?

File: codeinterpreterapi/agents/functions_agent.py [hunk 5]
```diff
@@ -16,22 +16,13 @@
...
-    AgentAction,
-    AgentFinish,
-    AIMessage,
-    BaseMessage,
-    BasePromptTemplate,
-    FunctionMessage,
-    OutputParserException,
-    SystemMessage,
-)
+from langchain.prompts.chat import (BaseMessagePromptTemplate,
```
Comment (shroominic) [inline codeinterpreterapi/agents/functions_agent.py:19]: and why this?

File: codeinterpreterapi/utils/parser.py [hunk 6]
```diff
@@ -39,7 +39,8 @@ def _type(self) -> str:
 
 class CodeChatAgentOutputParser(AgentOutputParser):
     def get_format_instructions(self) -> str:
-        from langchain.agents.conversational_chat.prompt import FORMAT_INSTRUCTIONS
+        from langchain.agents.conversational_chat.prompt import \
```
Comment (shroominic) [inline codeinterpreterapi/utils/parser.py:42]: Also this is not needed

File: codeinterpreterapi/prompts/remove_dl_link.py (hunk 4 above)
Comment (ducanhdt) [reply codeinterpreterapi/prompts/remove_dl_link.py:1]: I think this change was caused by black auto-format

**Comment by ducanhdt** [comment]
I run the precommit in the local machine and found that black and isort try to fix each other so I only can pass only one.

**Comment by shroominic** [comment]
Hey @ducanhdt i just forked your PR (#77) and made some improvements :)

**Comment by shroominic** [comment]
For me when I run pre-commit isort and black automatically change the code to resolve the issues
